ledger-core/conveyor/smachine/perf/perf_test.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package perf
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/smachine"
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/sworker"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/atomickit"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/synckit"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/throw"
+)
+
+func BenchmarkLoops(b *testing.B) {
+	const workCost = 0 //
+	b.Run("Full Loops", func(b *testing.B) {
+		const shortLoops = false
+		runBench(b, 1, 0, workCost, shortLoops, false)
+		runBench(b, 10, 0, workCost, shortLoops, false)
+		runBench(b, 1000, 0, workCost, shortLoops, false)
+		runBench(b, 100_000, 0, workCost, shortLoops, false)
+	})
+
+	b.Run("Full Loops vs Idle", func(b *testing.B) {
+		const shortLoops = false
+		runBench(b, 1000, 0, workCost, shortLoops, false)
+		runBench(b, 1000, 10, workCost, shortLoops, false)
+		runBench(b, 1000, 1000, workCost, shortLoops, false)
+		runBench(b, 1000, 100_000, workCost, shortLoops, false)
+//		runBench(b, 1000, 10_000_000, workCost, shortLoops, false)
+	})
+
+	b.Run("Short Loops", func(b *testing.B) {
+		const shortLoops = true
+		runBench(b, 1, 0, workCost, shortLoops, false)
+		runBench(b, 10, 0, workCost, shortLoops, false)
+		runBench(b, 1000, 0, workCost, shortLoops, false)
+		runBench(b, 100_000, 0, workCost, shortLoops, false)
+	})
+}
+
+func runBench(b *testing.B, activeCount, idleCount, workLoop int, shortLoop, boost bool, ) {
+	name := ""
+	if idleCount == 0 {
+		name = fmt.Sprintf("%6d SM", activeCount)
+	} else {
+		name = fmt.Sprintf("%d of %d SM", activeCount, activeCount + idleCount)
+	}
+	const scanCountLimit = math.MaxUint32
+
+	b.Run(name, func(b *testing.B) {
+		workLoopsTotal := atomickit.Uint64{}
+		b.ReportAllocs()
+		runBenchSlotMachine(scanCountLimit, b.N, boost, func(m *smachine.SlotMachine) {
+			for i := idleCount; i > 0; i-- {
+				sm := &idleSM{ counter: &workLoopsTotal }
+				m.AddNew(context.Background(), sm, smachine.CreateDefaultValues{})
+			}
+			for i := activeCount; i > 0; i-- {
+				sm := &benchSM{ shortLoop: shortLoop, workLoop: workLoop, counter: &workLoopsTotal }
+				m.AddNew(context.Background(), sm, smachine.CreateDefaultValues{})
+			}
+			b.ResetTimer()
+		})
+		b.SetBytes(int64(workLoopsTotal.Load() / uint64(b.N)))
+	})
+}
+
+func runBenchSlotMachine(scanCountLimit int, cycleCount int, boost bool, initFn func(*smachine.SlotMachine)) {
+	boostDuration := 10 * time.Millisecond
+	if !boost {
+		boostDuration = 0
+	}
+
+	signal := synckit.NewVersionedSignal()
+	m := smachine.NewSlotMachine(smachine.SlotMachineConfig{
```

Comment:
defer m.Stop() :octocat:

Reply:
not needed here - because there is no worker routine